File: python/examples/structural_analysis.py
#!/usr/bin/env python3
# Copyright (c) 2024-2026 Elias Bachaalany
# SPDX-License-Identifier: MPL-2.0
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# structural_analysis: Analyze structural properties of functions.
#
# Usage: python structural_analysis.py [host_url]
#
# Finds a non-trivial function, queries its switch tables, dominator tree,
# post-dominator tree, and natural loops, then decompiles it and inspects
# the token stream.

import sys
from collections import Counter

import libghidra as ghidra


def main() -> None:
    url = sys.argv[1] if len(sys.argv) >= 2 else "http://127.0.0.1:18080"
    client = ghidra.connect(url)

    # 1. Verify connection
    try:
        status = client.get_status()
    except ghidra.GhidraError as e:
        print(f"Cannot reach host at {url}: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"Connected: {status.service_name} v{status.service_version}")

    # 2. Find a non-trivial function (size > 0)
    print("\n--- Searching for a non-trivial function (size > 0) ---")
    try:
        funcs = client.list_functions(limit=100)
    except ghidra.GhidraError as e:
        print(f"ListFunctions failed: {e}", file=sys.stderr)
        sys.exit(1)

    target = None
    for f in funcs.functions:
        if f.size > 0:
            target = f
            break

    if target is None:
        print("No function with size > 0 found.", file=sys.stderr)
        sys.exit(1)

    addr = target.entry_address
    print(f"Selected: {target.name} at 0x{addr:x} ({target.size} bytes)")

    # 3. Switch tables
    print(f"\n--- Switch tables for {target.name} ---")
    try:
        sw_resp = client.list_switch_tables(range_start=addr, range_end=addr)
        switch_tables = sw_resp.switch_tables
    except ghidra.GhidraError as e:
        print(f"ListSwitchTables failed: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"Total switch tables: {len(switch_tables)}")
    for st in switch_tables:
        print(f"  Switch at 0x{st.switch_address:x}  ({st.case_count} cases,"
              f" default=0x{st.default_address:x})")
        for case in st.cases:
            print(f"    case {case.value} -> 0x{case.target_address:x}")

    # 4. Dominator tree
    print(f"\n--- Dominators for {target.name} ---")
    try:
        dom_resp = client.list_dominators(range_start=addr, range_end=addr)
        dominators = dom_resp.dominators
    except ghidra.GhidraError as e:
        print(f"ListDominators failed: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"Total dominator records: {len(dominators)}")
    for d in dominators:
        entry_marker = " (entry)" if d.is_entry else ""
        print(f"  0x{d.block_address:x}  idom=0x{d.idom_address:x}"
              f"  depth={d.depth}{entry_marker}")

    # 5. Post-dominator tree
    print(f"\n--- Post-dominators for {target.name} ---")
    try:
        pdom_resp = client.list_post_dominators(range_start=addr, range_end=addr)
        post_dominators = pdom_resp.post_dominators
    except ghidra.GhidraError as e:
        print(f"ListPostDominators failed: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"Total post-dominator records: {len(post_dominators)}")
    for p in post_dominators:
        exit_marker = " (exit)" if p.is_exit else ""
        print(f"  0x{p.block_address:x}  ipdom=0x{p.ipdom_address:x}"
              f"  depth={p.depth}{exit_marker}")

    # 6. Natural loops
    print(f"\n--- Loops for {target.name} ---")
    try:
        loops_resp = client.list_loops(range_start=addr, range_end=addr)
        loops = loops_resp.loops
    except ghidra.GhidraError as e:
        print(f"ListLoops failed: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"Total loops: {len(loops)}")
    for lp in loops:
        print(f"  header=0x{lp.header_address:x}  back_edge=0x{lp.back_edge_source:x}"
              f"  kind={lp.loop_kind}  blocks={lp.block_count}  depth={lp.depth}")

    # 7. Decompile and inspect tokens
    print(f"\n--- Decompilation tokens for {target.name} ---")
    try:
        decomp_resp = client.get_decompilation(address=addr)
    except ghidra.GhidraError as e:
        print(f"GetDecompilation failed: {e}", file=sys.stderr)
        sys.exit(1)

    decomp = decomp_resp.decompilation
    if decomp is None:
        print("No decompilation returned.")
    else:
        print(f"Function: {decomp.function_name}")
        print(f"Prototype: {decomp.prototype}")
        print(f"Completed: {decomp.completed}")
        if decomp.error_message:
            print(f"Error: {decomp.error_message}")

        tokens = decomp.tokens
        print(f"\nTotal tokens: {len(tokens)}")

        # Token kind distribution
        kind_counts = Counter(t.kind.name for t in tokens)
        print("\nToken kinds:")
        for kind, count in kind_counts.most_common():
            print(f"  {kind:12s}  {count}")

        # Show first 20 tokens with detail
        preview = tokens[:20]
        print(f"\nFirst {len(preview)} tokens:")
        for t in preview:
            parts = [f"kind={t.kind.name}"]
            if t.var_name:
                parts.append(f"var={t.var_name}")
            if t.var_type:
                parts.append(f"type={t.var_type}")
            if t.var_storage:
                parts.append(f"storage={t.var_storage}")
            detail = ", ".join(parts)
            print(f"  L{t.line_number}:{t.column_offset:<4d} {t.text!r:20s}  ({detail})")

    # 8. Summary
    print(f"\n--- Summary ---")
    print(f"  Function:       {target.name}")
    print(f"  Address:        0x{addr:x}")
    print(f"  Size:           {target.size} bytes")
    print(f"  Switch tables:  {len(switch_tables)}")
    print(f"  Dominators:     {len(dominators)}")
    print(f"  Post-dominators:{len(post_dominators)}")
    print(f"  Loops:          {len(loops)}")
    if decomp is not None:
        print(f"  Decomp tokens:  {len(decomp.tokens)}")
        print(f"  Decomp lines:   {decomp.pseudocode.count(chr(10)) + 1}")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: rust/examples/structural_analysis.rs
// Copyright (c) 2024-2026 Elias Bachaalany
// SPDX-License-Identifier: MPL-2.0
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// structural_analysis: Demonstrate structural analysis RPCs (switch tables, dominators,
// post-dominators, loops) and decompile token inspection.
//
// Usage: structural_analysis [host_url]

use libghidra as ghidra;
use std::collections::HashMap;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let url = args.get(1).map_or("http://127.0.0.1:18080", |s| s.as_str());
    let client = ghidra::connect(url);

    // 1. Get server status
    let status = client.get_status().unwrap_or_else(|e| {
        eprintln!("get_status failed: {e}");
        std::process::exit(1);
    });
    println!(
        "Connected to {} v{} (mode={})",
        status.service_name, status.service_version, status.host_mode
    );

    // 2. List functions and pick the first non-trivial one (size > 64 bytes)
    let funcs = client
        .list_functions(0, u64::MAX, 200, 0)
        .unwrap_or_else(|e| {
            eprintln!("list_functions failed: {e}");
            std::process::exit(1);
        });
    let target = funcs
        .functions
        .iter()
        .find(|f| f.size > 64)
        .unwrap_or_else(|| {
            eprintln!("No non-trivial function found (size > 64 bytes)");
            std::process::exit(1);
        });
    println!(
        "\nSelected function: {} at 0x{:x} ({} bytes)",
        target.name, target.entry_address, target.size
    );

    let range_start = target.entry_address;
    let range_end = target.entry_address + target.size;

    // 3. Switch tables
    let switch_resp = client
        .list_switch_tables(range_start, range_end, 100, 0)
        .unwrap_or_else(|e| {
            eprintln!("list_switch_tables failed: {e}");
            std::process::exit(1);
        });
    println!(
        "\nSwitch tables ({} found):",
        switch_resp.switch_tables.len()
    );
    for st in &switch_resp.switch_tables {
        println!(
            "  switch at 0x{:x}: {} cases, default -> 0x{:x}",
            st.switch_address, st.case_count, st.default_address
        );
        for c in &st.cases {
            println!("    case {} -> 0x{:x}", c.value, c.target_address);
        }
    }

    // 4. Dominators (immediate dominator tree)
    let dom_resp = client
        .list_dominators(range_start, range_end, 1000, 0)
        .unwrap_or_else(|e| {
            eprintln!("list_dominators failed: {e}");
            std::process::exit(1);
        });
    println!("\nDominators ({} blocks):", dom_resp.dominators.len());
    for d in &dom_resp.dominators {
        if d.is_entry {
            println!("  0x{:x} [ENTRY] depth={}", d.block_address, d.depth);
        } else {
            println!(
                "  0x{:x} idom=0x{:x} depth={}",
                d.block_address, d.idom_address, d.depth
            );
        }
    }

    // Build dominator tree children map for summary
    let mut dom_children: HashMap<u64, Vec<u64>> = HashMap::new();
    for d in &dom_resp.dominators {
        if !d.is_entry {
            dom_children
                .entry(d.idom_address)
                .or_default()
                .push(d.block_address);
        }
    }
    let max_dom_depth = dom_resp
        .dominators
        .iter()
        .map(|d| d.depth)
        .max()
        .unwrap_or(0);
    println!("  Max dominator tree depth: {max_dom_depth}");

    // 5. Post-dominators (immediate post-dominator tree)
    let pdom_resp = client
        .list_post_dominators(range_start, range_end, 1000, 0)
        .unwrap_or_else(|e| {
            eprintln!("list_post_dominators failed: {e}");
            std::process::exit(1);
        });
    println!(
        "\nPost-dominators ({} blocks):",
        pdom_resp.post_dominators.len()
    );
    for pd in &pdom_resp.post_dominators {
        if pd.is_exit {
            println!("  0x{:x} [EXIT] depth={}", pd.block_address, pd.depth);
        } else {
            println!(
                "  0x{:x} ipdom=0x{:x} depth={}",
                pd.block_address, pd.ipdom_address, pd.depth
            );
        }
    }
    let max_pdom_depth = pdom_resp
        .post_dominators
        .iter()
        .map(|pd| pd.depth)
        .max()
        .unwrap_or(0);
    println!("  Max post-dominator tree depth: {max_pdom_depth}");

    // 6. Loops (natural loops detected via back edges)
    let loops_resp = client
        .list_loops(range_start, range_end, 100, 0)
        .unwrap_or_else(|e| {
            eprintln!("list_loops failed: {e}");
            std::process::exit(1);
        });
    println!("\nLoops ({} found):", loops_resp.loops.len());
    for lp in &loops_resp.loops {
        println!(
            "  header=0x{:x} back_edge_from=0x{:x} kind={} blocks={} depth={}",
            lp.header_address, lp.back_edge_source, lp.loop_kind, lp.block_count, lp.depth
        );
    }

    // 7. Decompile the function and inspect tokens
    let decomp_resp = client
        .get_decompilation(target.entry_address, 30000)
        .unwrap_or_else(|e| {
            eprintln!("get_decompilation failed: {e}");
            std::process::exit(1);
        });
    let decomp = decomp_resp.decompilation.unwrap_or_else(|| {
        eprintln!("No decompilation returned");
        std::process::exit(1);
    });
    println!("\nDecompilation of '{}':", decomp.function_name);
    println!("  Prototype: {}", decomp.prototype);
    println!("  Completed: {}", decomp.completed);
    println!("  Locals: {}", decomp.locals.len());
    println!("  Tokens: {}", decomp.tokens.len());

    // Token kind histogram
    let mut kind_counts: HashMap<String, usize> = HashMap::new();
    for tok in &decomp.tokens {
        *kind_counts.entry(format!("{:?}", tok.kind)).or_default() += 1;
    }
    println!("\n  Token kind breakdown:");
    let mut sorted_kinds: Vec<_> = kind_counts.into_iter().collect();
    sorted_kinds.sort_by(|a, b| b.1.cmp(&a.1));
    for (kind, count) in &sorted_kinds {
        println!("    {kind:<16} {count}");
    }

    // Show first 20 tokens as a sample
    let sample_count = decomp.tokens.len().min(20);
    println!("\n  First {sample_count} tokens:");
    for tok in decomp.tokens.iter().take(sample_count) {
        let extra = if !tok.var_name.is_empty() {
            format!(" (var={})", tok.var_name)
        } else {
            String::new()
        };
        println!(
            "    L{}:C{} {:?} {:?}{}",
            tok.line_number, tok.column_offset, tok.kind, tok.text, extra
        );
    }

    // 8. Summary
    println!(
        "\n--- Structural Analysis Summary for '{}' ---",
        target.name
    );
    println!("  Function size:        {} bytes", target.size);
    println!(
        "  Switch tables:        {}",
        switch_resp.switch_tables.len()
    );
    println!(
        "  Total switch cases:   {}",
        switch_resp
            .switch_tables
            .iter()
            .map(|s| s.case_count)
            .sum::<u32>()
    );
    println!("  Dominator nodes:      {}", dom_resp.dominators.len());
    println!("  Max dom depth:        {max_dom_depth}");
    println!(
        "  Post-dominator nodes: {}",
        pdom_resp.post_dominators.len()
    );
    println!("  Max pdom depth:       {max_pdom_depth}");
    println!("  Natural loops:        {}", loops_resp.loops.len());
    println!("  Decompile tokens:     {}", decomp.tokens.len());
    println!("  Decompile locals:     {}", decomp.locals.len());
}
